src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+    #   operator precedence schedule.
+    # Mainly this effects rmul, eg:
+    # np_ary @ diag_mat
+    __array_ufunc__ = None
+
+    def __init__(self, data, dtype=None):
+        """
+        Instantiate a `DiagMatrix` with Numpy `data` shaped (...., self.count),
+        where `self.count` is the length of one diagonal vector.
+
+        Slower axes (if present) are taken to be stack axes.
+
+        :param data: Diagonal matrix entries.
+        :param dtype: Datatype. Default of `None` will attempt
+            passthrough of `data.dtype`.  When explicitly provided, will
+            attempt casting `data` as needed.
+        :return: `DiagMatrix` instance.
+        """
+
+        # Assign the datatype.
+        if dtype is None:
+            dtype = data.dtype
+        self.dtype = np.dtype(dtype)
+
+        # Assign the `data`
+        self._data = data.astype(self.dtype, copy=False)
+
+        # Assign shapes from `data`
+        self.count = self._data.shape[-1]
+        self.stack_shape = self._data.shape[:-1]
+        self.shape = self._data.shape
+        # Total number of stack elements
+        self.size = np.prod(self.stack_shape)
+
+        # Numpy interop
+        # https://numpy.org/devdocs/user/basics.interoperability.html#the-array-interface-protocol
+        self.__array_interface__ = self.asnumpy().__array_interface__
+        self.__array__ = self.asnumpy()
+
+    def stack_reshape(self, *args):
+        """
+        Reshape the stack axis.
+
+        :*args: Integer(s) or tuple describing the intended shape.
+
+        :returns: `DiagMatrix` instance.
+        """
+
+        # If we're passed a tuple, use that
+        if len(args) == 1 and isinstance(args[0], tuple):
+            shape = args[0]
+        else:
+            # Otherwise use the variadic args
+            shape = args
+
+        # Sanity check the size
+        if shape != (-1,) and np.prod(shape) != self.size:
+            raise ValueError(
+                f"Number of images {self.size} cannot be reshaped to {shape}."
+            )
+
+        return DiagMatrix(self._data.reshape(*shape, self._data.shape[-1]))
+
+    def asnumpy(self):
+        """
+        Return data as Numpy array.
+
+        Note this is a read-only view.
+
+        :return: Numpy array of `self.dtype`.
+        """
+
+        view = self._data.view()
+        view.flags.writeable = False
+        return view
+
+    def __repr__(self):
+        """
+        String represention describing instance.
+        """
+        return "DiagMatrix({}, {})".format(repr(self._data), repr(self.dtype))
+
+    def copy(self):
+        """
+        Returns new `DiagMatrix` which is a copy of `self`.
+
+        :return `DiagMatrix` like self
+        """
+
+        return DiagMatrix(self._data.copy())
+
+    def __getitem__(self, key):
+        """
+        Convenience wrapper, getter on self._data.
+        """
+
+        return self._data[key]
+
+    def __len__(self):
```

Comment:
If we want this to emulate the array behavior, shouldn't this be `self.stack_shape[0]`?

Reply:
Yeah, I think I was just going too fast there. Updates and changed the test. Thanks.